packages/block-editor/src/components/block-list/use-block-props/use-block-refs.js

```diff
@@ -37,30 +36,35 @@ export function useBlockRefProvider( clientId ) {
 }
 
 /**
- * Gets a ref pointing to the current block element. Continues to return the same
- * stable ref object even if the `clientId` argument changes. This hook is not
- * reactive, i.e., it won't trigger a rerender of the calling component if the
- * ref value changes. For reactive use cases there is the `useBlockElement` hook.
+ * Tracks the DOM element for the block identified by `clientId` and assigns it to the `ref`
+ * whenever it changes.
  *
- * @param {string} clientId The client ID to get a ref for.
- *
- * @return {RefObject} A ref containing the element.
+ * @param {string} clientId The client ID to track.
+ * @param {Ref}    ref      The ref object/callback to assign to.
  */
-function useBlockRef( clientId ) {
+export function useBlockElementRef( clientId, ref ) {
 	const { refsMap } = useContext( BlockRefs );
-	const latestClientId = useRef();
-	latestClientId.current = clientId;
-
-	// Always return an object, even if no ref exists for a given client ID, so
-	// that `current` works at a later point.
-	return useMemo(
-		() => ( {
-			get current() {
-				return refsMap.get( latestClientId.current ) ?? null;
-			},
-		} ),
-		[ refsMap ]
+	const setRef = useCallback(
+		( el ) => {
+			if ( typeof ref === 'function' ) {
+				ref( el );
+			} else if ( ref ) {
+				ref.current = el;
+			}
+		},
+		[ ref ]
```

Comment:
Could we not avoid the `useCallback` here by moving the function within the `useLayoutEffect` that follows?

Reply:
Or `setRef( ref, element )` could even be outside the hook

Reply:
Haha, there's actually also `assignRef` in `use-merge-refs`

Reply:
Not that we can use it here 

https://github.com/WordPress/gutenberg/blob/47bde1a7696ceae94557dde3fa57cde3cfc9d84e/packages/compose/src/hooks/use-merge-refs/index.js#L18-L27

Reply:
Here we need to ensure that when the `ref` parameter in `useBlockElementRef( clientId, ref )` changes, the new version is called immediately. That's what React also does. When a different `ref` is passed to `<div ref={ ref } />`, then even though the `div` element is still the same, it will call the old ref with `null` and the new ref with the element.

Reply:
> Not that we can use it here, it's a different package

Interesting, I didn't know about that. The `ref.hasOwnProperty( 'current' )` check is nice, but not even React bothers to do it. It checks `ref !== null`, `typeof ref === 'function'`, and then if `ref.current` assignment leads to a crash, it will crash, it's you fault.

Reply:
> The layout effect will still be called when adding ref as a dependency?

Now I get it, of course you're right, fixing in 26e153a3d2f. It's a nice simplification.